File: server/src/routes/orgClassifications.js
const express = require('express');
const { query, queryOne, execute, logAudit } = require('../db');
const { authRequired } = require('../middleware/auth');

const router = express.Router();

// GET all — public (used in registration dropdown)
router.get('/', async (req, res) => {
  try {
    const rows = await query('SELECT * FROM org_classifications ORDER BY sort_order, id');
    res.json(rows);
  } catch (err) { console.error(err); res.status(500).json({ error: 'Internal server error' }); }
});

// POST create
router.post('/', authRequired, async (req, res) => {
  try {
    if (req.user.role !== 'superadmin') return res.status(403).json({ error: 'Forbidden' });
    const { name } = req.body;
    if (!name?.trim()) return res.status(400).json({ error: 'Name is required' });
    const exists = await queryOne('SELECT id FROM org_classifications WHERE name = $1', [name.trim()]);
    if (exists) return res.status(409).json({ error: 'Classification already exists' });
    const maxRow = await queryOne('SELECT MAX(sort_order) m FROM org_classifications');
    const maxOrder = maxRow?.m || 0;
    const result = await execute('INSERT INTO org_classifications (name, is_enabled, is_system, sort_order) VALUES ($1, 1, 0, $2) RETURNING id', [name.trim(), maxOrder + 1]);
    const row = await queryOne('SELECT * FROM org_classifications WHERE id = $1', [result.rows[0].id]);
    await logAudit({ user: req.user, action: 'Org classification added', entity: 'org_classification', entityId: row.id, detail: name.trim(), ip: req.ip });
    res.status(201).json(row);
  } catch (err) { console.error(err); res.status(500).json({ error: 'Internal server error' }); }
});

// PATCH enable/disable
router.patch('/:id/status', authRequired, async (req, res) => {
  try {
    if (req.user.role !== 'superadmin') return res.status(403).json({ error: 'Forbidden' });
    const { is_enabled } = req.body;
    const row = await queryOne('SELECT * FROM org_classifications WHERE id = $1', [req.params.id]);
    if (!row) return res.status(404).json({ error: 'Not found' });
    await execute('UPDATE org_classifications SET is_enabled = $1 WHERE id = $2', [is_enabled ? 1 : 0, row.id]);
    await logAudit({ user: req.user, action: is_enabled ? 'Org classification enabled' : 'Org classification disabled', entity: 'org_classification', entityId: row.id, detail: row.name, ip: req.ip });
    res.json({ ...row, is_enabled: is_enabled ? 1 : 0 });
  } catch (err) { console.error(err); res.status(500).json({ error: 'Internal server error' }); }
});

// PATCH rename
router.patch('/:id', authRequired, async (req, res) => {
  try {
    if (req.user.role !== 'superadmin') return res.status(403).json({ error: 'Forbidden' });
    const { name } = req.body;
    if (!name?.trim()) return res.status(400).json({ error: 'Name is required' });
    const row = await queryOne('SELECT * FROM org_classifications WHERE id = $1', [req.params.id]);
    if (!row) return res.status(404).json({ error: 'Not found' });
    if (row.is_system) return res.status(400).json({ error: 'System classifications cannot be renamed' });
    const dup = await queryOne('SELECT id FROM org_classifications WHERE name = $1 AND id != $2', [name.trim(), row.id]);
    if (dup) return res.status(409).json({ error: 'Name already exists' });
    await execute('UPDATE org_classifications SET name = $1 WHERE id = $2', [name.trim(), row.id]);
    res.json({ ...row, name: name.trim() });
  } catch (err) { console.error(err); res.status(500).json({ error: 'Internal server error' }); }
});

// DELETE
router.delete('/:id', authRequired, async (req, res) => {
  try {
    if (req.user.role !== 'superadmin') return res.status(403).json({ error: 'Forbidden' });
    const row = await queryOne('SELECT * FROM org_classifications WHERE id = $1', [req.params.id]);
    if (!row) return res.status(404).json({ error: 'Not found' });
    if (row.is_system) return res.status(400).json({ error: 'System classifications cannot be deleted' });
    await execute('DELETE FROM org_classifications WHERE id = $1', [row.id]);
    await logAudit({ user: req.user, action: 'Org classification deleted', entity: 'org_classification', entityId: row.id, detail: row.name, ip: req.ip });
    res.json({ success: true });
  } catch (err) { console.error(err); res.status(500).json({ error: 'Internal server error' }); }
});

module.exports = router;

File: server-python/routers/org_classifications.py
from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel
from database import query, query_one, execute, execute_returning, log_audit
from auth import auth_required, require_role

router = APIRouter(prefix="/api/org-classifications", tags=["org-classifications"])

DEFAULTS = [
    "Private Limited Company", "Public Limited Company", "Partnership Firm",
    "Sole Proprietorship", "LLP (Limited Liability Partnership)",
    "Society / Trust / NGO", "Government Institution", "Autonomous Body",
]


async def _seed_defaults():
    count_row = await query_one("SELECT COUNT(*) c FROM org_classifications")
    if count_row and int(count_row["c"] or 0) == 0:
        for i, name in enumerate(DEFAULTS):
            await execute(
                "INSERT INTO org_classifications (name, is_enabled, is_system, sort_order) VALUES ($1, 1, 1, $2) ON CONFLICT (name) DO NOTHING",
                [name, i + 1],
            )


@router.get("")
async def list_classifications():
    await _seed_defaults()
    rows = await query("SELECT * FROM org_classifications ORDER BY sort_order, id")
    return rows


class ClassificationIn(BaseModel):
    name: str


@router.post("")
async def create_classification(body: ClassificationIn, user: dict = Depends(require_role("superadmin"))):
    name = body.name.strip()
    if not name:
        raise HTTPException(400, "Name is required")
    existing = await query_one("SELECT id FROM org_classifications WHERE name=$1", [name])
    if existing:
        raise HTTPException(400, "Already exists")
    max_row = await query_one("SELECT MAX(sort_order) m FROM org_classifications")
    max_order = int(max_row["m"] or 0) if max_row else 0
    row = await execute_returning(
        "INSERT INTO org_classifications (name, is_enabled, is_system, sort_order) VALUES ($1, 1, 0, $2) RETURNING *",
        [name, max_order + 1],
    )
    await log_audit(user, "Org classification added", "org_classification", row["id"], name)
    return row


class StatusIn(BaseModel):
    is_enabled: bool


@router.patch("/{id}/status")
async def set_status(id: int, body: StatusIn, user: dict = Depends(require_role("superadmin"))):
    row = await query_one("SELECT * FROM org_classifications WHERE id=$1", [id])
    if not row:
        raise HTTPException(404, "Not found")
    await execute("UPDATE org_classifications SET is_enabled=$1 WHERE id=$2", [1 if body.is_enabled else 0, id])
    action = "Org classification enabled" if body.is_enabled else "Org classification disabled"
    await log_audit(user, action, "org_classification", id, row["name"])
    return {**row, "is_enabled": 1 if body.is_enabled else 0}


class RenameIn(BaseModel):
    name: str


@router.put("/{id}")
async def rename_classification(id: int, body: RenameIn, user: dict = Depends(require_role("superadmin"))):
    name = body.name.strip()
    if not name:
        raise HTTPException(400, "Name is required")
    row = await query_one("SELECT * FROM org_classifications WHERE id=$1", [id])
    if not row:
        raise HTTPException(404, "Not found")
    dup = await query_one("SELECT id FROM org_classifications WHERE name=$1 AND id!=$2", [name, id])
    if dup:
        raise HTTPException(400, "Name already in use")
    await execute("UPDATE org_classifications SET name=$1 WHERE id=$2", [name, id])
    await log_audit(user, "Org classification renamed", "org_classification", id, f"{row['name']} → {name}")
    return {**row, "name": name}


@router.delete("/{id}")
async def delete_classification(id: int, user: dict = Depends(require_role("superadmin"))):
    row = await query_one("SELECT * FROM org_classifications WHERE id=$1", [id])
    if not row:
        raise HTTPException(404, "Not found")
    if row.get("is_system"):
        raise HTTPException(400, "Cannot delete a system classification")
    await execute("DELETE FROM org_classifications WHERE id=$1", [id])
    await log_audit(user, "Org classification deleted", "org_classification", id, row["name"])
    return {"ok": True}
